KEP-4322: Cluster Inventory API (#4323)
kubernetes/enhancements: keps/sig-multicluster/4322-cluster-inventory/README.md

- **Cluster Manager**: An entity that creates the cluster inventory API 
  object per member cluster, and keeps their status up-to-date. Each
  cluster manager SHOULD be identified with a unique name. Each cluster
  inventory resource SHOULD be managed by only one cluster manager. Examples
  of cluster manager are projects like OCM, Karmada, Clusternet or Azure
  fleet manager.

- **Cluster Inventory Consumer**: the person running the cluster managers
  or the person developing extensions for cluster managers for the purpose of
  workload distribution, operation management etc.

- **Member Cluster**: A kubernetes cluster that is managed by the cluster
  manager. A cluster manager SHOULD have sufficient permission to access
  the member cluster to fetch the information so it can update the status
  of the cluster inventory API resource.

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1: Multicluster Workload Distribution

> In this scenario, a standard API acts as a foundation for multicluster

Reviewer: Did you mean to say "unified" API (which is how this new API is referred to elsewhere)?

Author: changed to unified

scheduling decisions across various clusters. This means that the
API will provide a single point of contact to retrieve information that can be
used to make informed scheduling decisions for workloads across multiple
clusters. For instance, workload distribution tools like GitOps or
[Work API](https://github.com/kubernetes-sigs/work-api) can leverage this
API to make informed decisions about which cluster is best suited to handle
a particular workload.

Examples of the multicluster scheduling includes:

- As a user I want to run a workload on an EKS cluster that resides in us-east-1.
  I want to submit my workload if and only if a cluster satisfies that constraint.
- As a user I want to run a workload on a cluster that has certain CRDs installed.

> - As a user I want to deploy a workload to my on-prem cluster group.

Reviewer: cluster group means something different to each reader, as we're not implying grouping in this KEP, it might be better to just reference "...to my on-prem cluster."

- As a user I want to run a workload close to the data source or end-users.

> - As a user I want to run a workload to the less busy cluster.

Reviewer 2: Should we add a Cluster API story here? Cluster API users are a natural target customer for this scenario, as the nominal use case is a multicluster one.

Reviewer 3: Yes, I totally agree.

should be the same as the `spec.clusterManager.name`. Different cluster
manager must set a different value of `spec.clusterManager.name` when
the cluster is created.

### Spec

#### Display name

It is a human-readable name of the cluster set by the consumer of the
cluster.

#### Cluster Manager

An immutable field set by a cluster manager when this cluster resources
is created by it. Each cluster manager instance should set a different
values to this field.

In addition, a predefined label with key "x-k8s.io/cluster-manager" needs
to be added by the cluster manager upon creation. The value of the label
MUST be the same as the name of the cluster manager. The purpose of this
label is to make filter clusters from different cluster managers easier.

### Status

#### Version

Kubernetes version of the cluster

Versions of the kubernetes can let consumers understand the capability
of the kubernetes, such as what API is supported.

With recent conversations about kube-apiserver and enabled featureset version,
it is possible to incorporate other version relating to the cluster, such as
minimum kubelet version, maximum kubelet version, and enabled featureset version.

> #### Properties

Reviewer: Is the plan to make this enumerated (defined by the API or free form, where each Cluster Manager would have its own defined list of properties?

Reviewer: Actually, reading the conditions, having some predefined properties, and then allow custom properties from the Cluster Manager, would probably help.

  this condition.
- Schedulable: define if the cluster is eligible to be scheduled the workload. By
  default, if at lease one node in the cluster is schedulable, the cluster will be set
  as schedulable. In the future, we would consider a spec field to define the criteria
  of schedulable status.

## API Example

```yaml
apiVersion: multicluster.x-k8s.io/v1alpha1
kind: ClusterInventory
metadata:
 name: generated-cluster-name
 labels:
   x-k8s.io/cluster-manager: some-cluster-manager
spec:
  displayName: some-cluster
  clusterManager:
    name: some-cluster-manager
status:
 version:
   kubernetes: 1.28.0
 properties:
   - name: clusterset.k8s.io
     value: some-clusterset
   - name: location
     value: apac
 conditions:
   - type: ControlPlaneHealthy
     status: True
     lastTransitionTime: "2023-05-08T07:56:55Z"
     message: ""
   - type: AllNodesHealthy
     status: True
     lastTransitionTime: "2023-05-08T07:57:55Z"
     message: ""
   - type: Joined
     status: True
     lastTransitionTime: "2023-05-08T07:58:55Z"
     message: ""
   - type: Schedulable
     status: True
     lastTransitionTime: "2023-05-08T07:58:55Z"
     message: ""
```

### Scalability implication

> The API should provide summarized metadata of the cluster and relatively "static" cluster status.

Reviewer 4: perhaps adding some thoughts in terms of burst+qps update rates?  Burst of 10 updates (to handle early join or sudden storms), with a qps/refill of one every 30s (10s, 60s?) ?

I'm less concerned with specific values and more interested in those coordinating having a shared sense of reactivity.

Author: that makes sense, updated.

> Dynamic data, e.g. cluster resource usage, should not be included in this API give it will bring

Reviewer: Should the bullet up above in the Design section where there is a user question:

> Does the cluster have sufficient resources?

Be adjusted to specifically imply that its Boolean or something less variable then the percentage or available value

Author: The wording of the scalability section is updated.

* Offering functionalities related to multi-cluster orchestration.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

the API proposed by this KEP aims to
- Provide a reliable, consistent, and automated approach for any
  multi-cluster application (framework, toolset) to discover available
  clusters and take actions accordingly, in a way similar to service
  discovery works in a microservice architecture. Through the inventory,
  the application can query for a list of clusters to access, or watch
  for an ever-flowing stream of cluster lifecycle events which the
  application can act upon timely, such as auto-scaling, upgrades,
  failures, and connectivity issues.
- Provide a simple, clear interface for human operators to understand
  clusters under management.

### Terminology

- **Cluster Manager**: An entity that creates the cluster inventory API 

>   resources per member cluster, and keeps their status up-to-date. Each

Reviewer 5: Do you mean "objects" rather than "resources"?  The word "resource" is often used in the kubernetes codebase to refer to a kind of object rather than an individual.

Author: change to object

The target consumers of the API are users who manage the clusters and
other tools to understand the clusters concept from the API for
multicluster scheduling, workload distribution and cluster management.
The API aims to provide the information for the consumers to answer the
below questions:

* Is the cluster under management?
* How can I select a cluster?
* Is the cluster healthy?
* Does the cluster have certain capabilities or properties?
* Does the cluster have sufficient resources?

### Cluster Name

It is required that cluster name is unique for each cluster, and it
should also be unique among different providers (cluster manager). There
are two option to ensure it is uniqueness

#### Option 1

The `metadata.name` of the cluster should never be set upon creation.
Only `metadata.generateName` can be set.

#### Option 2

The `metadata.name` and `metadata.generateName` must have prefix which
should be the same as the `spec.clusterManager.name`. Different cluster
manager must set a different value of `spec.clusterManager.name` when
the cluster is created.

These two options can be achieved by a validating webhook or a
`ValidatingAdmissionPolicy` resource.

### Spec

> 

Reviewer 3: Hi @ArangoGutierrez so add a condition like this:
```
 conditions:
   - type: NodeSelectorsSatisfied
     status: True
     lastTransitionTime: "2023-05-08T07:57:55Z"
```
if the ClusterManager found more than 1 node that matches the nodeSelector? 
Could you give a more concrete example?
I agree we should start with a boolean type. The expectation is this boolean wouldn't flip flop too often but static for a while.

> 

Reviewer 6: Yes, but individually, `NodeSelectorsSatisfied` is plural and provides little info. 
```yaml
 conditions:
   - type: NodeSelectorSatisfied
     status: True
     message: NodeSelector XYZ
     lastTransitionTime: "2023-05-08T07:57:55Z"
```

> 

Author: Is this nodeSelector imply that any workload deploy to this cluster should be scheduled using this nodeSelector?
What does NodeSelectorsSatisfied indicate? all selected nodes are schedulable?
If we set an empty nodeSelector, It will select all nodes in the clusters, right?

> 

Reviewer 6: > Is this nodeSelector imply that any workload deploy to this cluster should be scheduled using this nodeSelector?

No

> What does NodeSelectorsSatisfied indicate? all selected nodes are schedulable?

that the target Cluster (of the ClusterInventory) can handle pods with that nodeSelector, (the resources or features exists) 

> If we set an empty nodeSelector, It will select all nodes in the clusters, right?

NodeSelector here is not for "selecting" nodes, but for querying if the Cluster has resources/features/labels of that type.

i.e: ClusterInventory says the cluster has enough CPU and Mem, but we also wanna know if the Cluster has a XYZ label indicating that some nodes have some extra configuration needed for a workload.

> 

Author: added a schedulable status condition. I am thinking we may put nodeSelector spec field as beta plan.

> #### Display name

Reviewer 6: Giving it more thought, Display name and ClusterManager should be part ObjectMeta.Name and ObjectMeta.Labels 
not part of the spec of ClusterInventory

Reviewer 6: The good thing about this approach, is that we could leverage  `corev1.NodeSelectorTerm` and it's matcher methods. https://github.com/kubernetes/api/blob/master/core/v1/types.go#L3166

Author: metadata.name would have more strict validation and immutable. Display name on the other hand can be more flexible. 

I feel like ClusterManager is needed such that different provider can know which cluster they need to care under the case that multiple providers run in the same control plane. And having it in the spec field could allow stronger validations.

  different scenarios. We will have multiple condition types to define
  healthiness:
  - ControlPlaneHealthy is to define whether the controlplane of the
    cluster is in the healthy state.
    * apiserver/healthz
    * controller-manager/healthz
    * scheduler/healthz
    * etcd/healthz
  - <<[UNRESOLVED]>> AllNodesHealthy is to define if the nodes in the cluster are in a
    healthy state. If one node is not healthy, the status of NodeHealthy
    is set to false with the message indicating details. (todo
    tolerance, it should be configurable)

  It would be useful to collect the healthiness of other subsystems in
  the cluster, e.g. network, dns, storage, or ingress. However, it is
  not easy to collect that information in a common way with different
  implementations of network or storage providers. We decide not to
  include other subsystems healthiness conditions in the initial phase.
- Joined: indicate the cluster is under management by the cluster manager.
  The status of the cluster SHOULD be updated by the cluster manager under
  this condition.

## API Example

```yaml
apiVersion: multicluster.x-k8s.io/v1alpha1

> kind: ClusterInventory

Reviewer 5: This Kind looks odd to me. In plain English, "inventory" is a _list of_ items, it is not an item in that list. Wouldn't something like `Cluster` be more appropriate here?

  resides in, resources a cluster provides, or latency to some external
  endpoints. A common inventory cluster API will give schedulers a
  standard to reason about clusters and help to foster the growth of
  this area.
* GitOps tools (ArgoCD, flux etc) are having the requirement to deploy
  workload to multiple clusters. They either need to build the cluster
  concept by themselves or understand cluster API from each different
  cluster management project. A common inventory cluster API can provide
  a thin compatible layer for different projects.
* Operation tools or customized external consumers: this API gives a
  common way for different clouds and vendors to define clusters,
  providing a vendor agnostic integration point for external tools.
* Cluster manager implementations themselves, for purposes such as
  grouping clusters into MCS API clustersets.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

* Establish a standardized inventory cluster API to represent clusters.
* Lay the groundwork for multi-cluster tooling by providing a
  foundational component.

> * Accommodate multiple implementations to encourage flexibility and

Reviewer 7: is this KEP also aiming to define a common interface similar to CRI/CSI?

Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

We try to summarize the most common properties that should be able to
represent a cluster and support the above use case at the minimum scope.

The target consumers of the API are users who manage the clusters and
other tools to understand the clusters concept from the API for
multicluster scheduling, workload distribution and cluster management.
The API aims to provide the information for the consumers to answer the
below questions:

* Is the cluster under management?
* How can I select a cluster?
* Is the cluster healthy?
* Does the cluster have certain capabilities or properties?
* Does the cluster have sufficient resources?

### Cluster Name

> It is required that cluster name is unique for each cluster, and it

Reviewer 8: The name should be unique within the inventory (group) and among different providers (cluster manager). The inventory could manage clusters created by multiple cloud providers and from different regions.  In your example, the cluster name will become "foobar-eks-west", "foobar-eks-east", "foobar-gke-west", "foobar-gke-east".

> It is required that cluster name is unique for each cluster, and it

Reviewer 2: I know it isn't mentioned in the doc now, but are we assuming that we will be reusing the `ObjectMeta` type here, and in which case should we explicitly mention a `Namespace` property? I think we want to allow for > 1 inventory to exist in a single reference implementation, and this seems the k8s-idiomatic way to accommodate that.

  different scenarios. We will have multiple condition types to define
  healthiness:
  - ControlPlaneHealthy is to define whether the controlplane of the
    cluster is in the healthy state.
    * apiserver/healthz
    * controller-manager/healthz
    * scheduler/healthz
    * etcd/healthz
  - <<[UNRESOLVED]>> AllNodesHealthy is to define if the nodes in the cluster are in a
    healthy state. If one node is not healthy, the status of NodeHealthy
    is set to false with the message indicating details. (todo
    tolerance, it should be configurable)
  It would be useful to collect the healthiness of other subsystems in
  the cluster, e.g. network, dns, storage, or ingress. However, it is
  not easy to collect that information in a common way with different
  implementations of network or storage providers. We decide not to
  include other subsystems healthiness conditions in the initial phase.
- Joined: indicate the cluster is under management by the cluster manager.
  The status of the cluster SHOULD be updated by the cluster manager under
  this condition.

## API Example

```yaml
apiVersion: multicluster.x-k8s.io/v1alpha1

> kind: InventoryCluster

Reviewer 2: Do we have a mention why we aren't able to re-use an existing standard API to represent a cluster. For example:

- https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/cluster_types.go#L489-L495

Reviewer 3: Right. Maybe we can add a mention to why consumers and providers prefer a more "neutral" API like this one over something like clusterapi which imo has a coupling with cluster provisioning.

Reviewer 2: To be fair we are fully qualifying the resource under the `multicluster.x-k8s.io/v1alpha1` group/version in this proposal, whereas CAPI is represented under `cluster.x-k8s.io/v1beta1`. So I don't think we're making any serious violation.

Sorry to be annoying but I'd probably place a vote for `ClusterRecord` instead of `InventoryCluster` or `ClusterInventory`. Or is there a better word than "record" to describe a unit in an inventory?

Reviewer 3: I am not a big fan of `ClusterRecord` or `ClusterProfile` or something similar if we are planning to have the API access info in there as well for phase 2. WDYT?

Reviewer 9: I think something like `ClusterProfile` or `ClusterRecord` would actually make a lot of sense for phase 2 needs too - semantically a "profile" and a "record" are both summary descriptions of an entity (the cluster in this case), which would both be written to by a user as a means of describing the entity (in Kubernetes APIs this would be through the `spec` section) or updated/categorized/reported on by a system or administrator (this would be a controller updating the `status` section).

I'll digress a bit about into a pedantic critique of specifically why I think the `InventoryCluster` term feels awkward - in many other Kubernetes APIs, the terminology pattern for combined camel-case APIs is for the last half of the term to be a generic "type" of resource, and the first half to be a means of qualifying the scope or role of the resource - examples include `DaemonSet`, `ReplicaSet`, `StatefulSet` and `RoleBinding`, `ClusterRoleBinding`. Following this pattern, the structure of the term `InventoryCluster` could imply that the object it represents is a cluster used exclusively for inventory purposes (think how https://open-cluster-management.io/concepts/architecture/#hub-spoke-architecture uses terms structured this way for "Hub Cluster" and "Managed Cluster"). Additionally, cluster-scoped resources typically follow the above-described pattern by using a `Cluster*` _prefixed_ term to indicate their scope (see `ClusterRoleBinding`, [`ClusterProperty`](https://github.com/kubernetes-sigs/about-api), etc), and I believe it would help users better understand the usage of this API if it followed terminology patterns more similar to existing APIs.

Reviewer 3: Right. We should follow convention and not use something that feels unnatural.
Let's gather some more feedback and hopefully come to a conclusion in the next SIG MC meeting.

Reviewer 10: @jackfrancis 

I think this API serves a different purpose from cluster-api. cluster-api describes how to construct a cluster. This API describes the cluster for whoever wants to use the cluster in a multi-cluster environment.

We should allow customers who don't use cluster API to provision their cluster to be able to describe a cluster in a multi-cluster environment.

Reviewer 2: The `Cluster` interface in Cluster API just _decribes a cluster_, similar to what is being designed here. I agree that the way that Cluster API describes a cluster (ref: https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/cluster_types.go#L45) is probably not general enough for re-use in other related concerns (which is unfortunate), including this one.

And certainly agree that we don't want to require Cluster API for the inventory use case, just wondering if we should add a note for objective observers who might wonder (with justification IMO!) "Why does the Kubernetes project define a Cluster many different ways?"

Reviewer 11: > The `Cluster` in Cluster API just _decribes a cluster_, similar to what is being designed here. I agree that the way that Cluster API describes a cluster (ref: https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/cluster_types.go#L45) is probably not general enough for re-use in other related concerns (which is unfortunate), including this one.

The Cluster API `Cluster` is more than that, isn't it? It describes the state of the cluster as the operator wants it to be. InventorCluster (or whatever we go with :) ) is closer to Cluster API's `Cluster.status`. To me this shows a mismatch - beyond just the fact that Cluster API shouldn't be required - the two resources are owned by different entities. Cluster API's `Cluster` is owned by the platform administrator provisioning clusters, the sig-mc `ClusterInventory` is owned by the cluster manager that created the cluster it represents.

Reviewer 9: > It seems that this is exactly what we are trying to achieve here. Why is that not an accurate description of the name?

Following the examples above, it could be inferred as describing to the _internal contents_ of the cluster, like "a Kubernetes cluster which only holds inventory records" as opposed to the _description_ of a specific cluster in a list of records.

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

* Provide a standard reference implementation.
* Define specific implementation details beyond general API behavior.
* Offering functionalities related to multi-cluster orchestration.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

the API proposed by this KEP aims to
- Provide a reliable, consistent, and automated approach for any
  multi-cluster application (framework, toolset) to discover available
  clusters and take actions accordingly, in a way similar to service
  discovery works in a microservice architecture. Through the inventory,
  the application can query for a list of clusters to access, or watch
  for an ever-flowing stream of cluster lifecycle events which the
  application can act upon timely, such as auto-scaling, upgrades,

>   failures, and connectivity issues.

Reviewer 11: I think we should consider this part carefully. Assuming this API evolves to capture more than basic cluster info (name, location), this data could be very noisy. What are the scaling implications of supporting watch on potentially high-churn cluster state across many clusters?

- Provide a simple, clear interface for human operators to understand
  clusters under management, and troubleshoot them in a streamlined

>   manner.

Reviewer 11: This API could simplify discovery and a high level understanding of state, but would it actually be part of troubleshooting?

Reviewer 6: Agree, maybe this is not a "troubleshooting" / Debugging API

Author: +1. removed the wording on debugging.

Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1: Multicluster Workload Distribution

In this scenario, a standard API acts as a foundation for multicluster
scheduling decisions across various clusters. This means that the
API will provide a single point of contact to retrieve information that can be
used to make informed scheduling decisions for workloads across multiple
clusters. For instance, workload distribution tools like GitOps or
[Work API](https://github.com/kubernetes-sigs/work-api) can leverage this
API to make informed decisions about which cluster is best suited to handle
a particular workload.

Examples of the multicluster scheduling includes:

- As a user I want to run a workload on an EKS cluster that resides in us-east-1.
  I want to submit my workload if and only if a cluster satisfies that constraint.
- As a user I want to run a workload on a cluster that has Kueue/Yunikorn enabled.
  The cluster level scheduling wants to submit a workload on a cluster that has the

>   shortest time to being picked up.

Reviewer 11: I'm still concerned about scheduling implications. There are a few dimensions to resource availability, and for this to be generally useful they need to be well understood. State also needs to be updated in realtime as stale data can drive bad decisions, or require verification, which can lessen the usefulness of having this likely high-churn data.

The other items in this list reflect intrinsic cluster properties (cluster location/platform) or low churn state (CRD installation) and seem like a safer place to start

Author: removed this use case, will add a separate section on it should not handle realtime data and the scalability implication.

- As a user I want to run a workload on a cluster that has certain CRDs installed.
- As a user I want to deploy a workload to my on-prem cluster group.
- As a user I want to run a workload close to the data source or end-users.
- As a user I want to run a workload to the less busy cluster.

#### Story 2: Operations and Management

For multi-cluster administrators, the unified API provides a
comprehensive view of all the clusters under their management. This
includes verifying their memberships, understanding their status,
capacity, and healthiness.

For instance, the API could provide real-time data about the CPU usage,
memory consumption, network latency, and other key performance

> indicators of each cluster. If a cluster is nearing its capacity or

Reviewer 11: Same concerns here - this kind of data can be very high traffic, what are the implications to scalability?

Is this API the right place for this sort of data? There are monitoring solutions that are built specifically to handle metrics like these, that really should accompany any production deployment already and support the sort of alerting we'd want here. Can we hand off to them?

Author: yes, removed this part and add resource based scheduling as beta plan.

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

We try to summarize the most common properties that should be able to
represent a cluster and support the above use case at the minimum scope.

The target consumers of the API are users who manage the clusters and
other tools to understand the clusters concept from the API for
multicluster scheduling, workload distribution and cluster management.
The API aims to provide the information for the consumers to answer the
below questions:

* Is the cluster under management?
* How can I select a cluster?
* Is the cluster healthy?
* Does the cluster have certain capabilities or properties?
* Does the cluster have sufficient resources?

### Cluster Name

It is required that cluster name is unique for each cluster, and it
should also be unique among different providers (cluster manager). There

> are two option to ensure it is uniqueness

Reviewer 11: So long as clusters reside in the same namespace (or aren't namespaced), name uniqueness is already ensured. Why do we need to force certain patterns? I assume every cluster manager already has some naming system, and it seems like it would be good to use whatever is already provided?

Reviewer 6: This has been a topic of debate on every call about this (Sig-MC meeting). I feel we should drop this from the initial v1alpha1 and see if the community needed this at the end for the next iteration of the API

Author: changed the wording here. It is provides the responsibility to ensure the name uniqueness. The proposal only provides two possible options but not mandatory.

#### Option 1

The `metadata.name` of the cluster should never be set upon creation.
Only `metadata.generateName` can be set.

#### Option 2

The `metadata.name` and `metadata.generateName` must have prefix which
should be the same as the `spec.clusterManager.name`. Different cluster
manager must set a different value of `spec.clusterManager.name` when
the cluster is created.

These two options can be achieved by a validating webhook or a
`ValidatingAdmissionPolicy` resource.

### Spec

#### Display name

It is a human-readable name of the cluster set by the consumer of the
cluster.

#### Cluster Manager

An immutable field set by a cluster manager when this cluster resources
is created by it. Each cluster manager instance should set a different
values to this field.

In addition, a predefined label with key "x-k8s.io/cluster-manager" needs
to be added by the cluster manager upon creation. The value of the label
MUST be the same as the name of the cluster manager. The purpose of this

> label is to make filter clusters from different cluster managers eaiser.

Reviewer 3: ```suggestion
label is to make filter clusters from different cluster managers easier.
```

### Status

#### Version

Kubernetes version of the cluster

Versions of the kubernetes can let consumers understand the capability
of the kubernetes, such as what API is supported.

With recent conversations about kube-apiserver and enabled featureset version,
it is possible to incorporate other version relating to the cluster, such as
minimum kubelet version, maximum kubelet version, and enabled featureset version.

#### Resources

Resources include capacity and allocatable, which is the sum of the
capacity/allocatable of nodes on the cluster. It is useful for users to
know the sizing of the cluster, and also schedulers to sort or bin pack

> the workload to clusters.

Reviewer 11: I brought this up in the doc, but capacity/allocatable needs definition. What are we showing? CPUs and type? GPUs? there are many resources and they are not fungible. Sum of CPU is not a useful metric in a cluster with mixed infra. Even with only one CPU type, there are many cases where it breaks down: e.g. 100 CPU free, but only 8 CPU/node and I'm trying to schedule a 32 CPU pod

Author: removed the resource field, and added the resource base scheduling as the beta plan

- A dummy controller and unit test to validate the CRD and client.

#### Beta

- Gather feedback from users during the Alpha stage to identify any
  issues, limitations, or areas for improvement. Address this feedback
  by making the necessary changes to the API and iterating on its design
  and functionality.
- The API should support the addition of scheduling features, such as:
  - Load balancing: Distribute workloads evenly across clusters based on
    their current load and capacity.
  - Affinity and anti-affinity rules: Allow users to define rules for
    placing workloads on specific clusters or ensuring that certain
    workloads are not placed on the same cluster.
  - Priority-based scheduling: Enable users to assign priorities to
    workloads, ensuring that higher-priority workloads are scheduled
    before lower-priority ones.
  - Resource-based scheduling: Schedule workloads based on the
    availability of specific resources, such as CPU, memory, or storage,
    in each cluster.
- The API should expose access information including but not limited to:
  - APIServer endpoint url of the member cluster.
  - Credential with limited access to the member cluster.

> - At least two providers and one consumers using cluster inventory API.

Reviewer 3: ```suggestion
- At least two providers and one consumer using cluster inventory API.
```

  endpoints. A common cluster inventory API will give schedulers a
  standard to reason about clusters and help to foster the growth of
  this area.
* GitOps tools (ArgoCD, flux etc) are having the requirement to deploy
  workload to multiple clusters. They either need to build the cluster
  concept by themselves or understand cluster API from each different
  cluster management project. A common cluster inventory API can provide
  a thin compatible layer for different projects.
* Operation tools or customized external consumers: this API gives a
  common way for different clouds and vendors to define clusters,
  providing a vendor agnostic integration point for external tools.
* Cluster manager implementations themselves, for purposes such as
  grouping clusters into MCS API clustersets.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

* Establish a standardized cluster inventory API to represent clusters.
* Lay the groundwork for multi-cluster tooling by providing a
  foundational component.
* Accommodate multiple implementations to encourage flexibility and
  adoption.

> * Allow for future extensions and new use cases by leaving room for

Author: added this as a goal

> * Allow for future extensions and new use cases by leaving room for

Reviewer: +1

> * Allow for future extensions and new use cases by leaving room for

Reviewer 7: +1

- **Cluster Manager**: An entity that creates the cluster inventory API 
  resources per member cluster, and keeps their status up-to-date. Each
  cluster manager SHOULD be identified with a unique name. Each cluster
  inventory resource SHOULD be managed by only one cluster manager. Examples
  of cluster manager are projects like OCM, Karmada, Clusternet or Azure
  fleet manager.

- **Cluster Inventory Consumer**: Users or third party tools that use
  cluster inventory API for the purpose of workload distribution, operation
  management etc. The consumer SHOULD not create this API resource.

- **Member Cluster**: A kubernetes cluster that is managed by the cluster
manager. A cluster manager SHOULD have sufficient permission to access
the member cluster to fetch the information so it can update the status
of the cluster inventory API resource.

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1: Multicluster Workload Distribution

> In this scenario, the unified API acts as a foundation for multicluster
> scheduling across various clusters. This means that the API will provide
> a standardized way to distribute workloads across multiple clusters. For
> instance, workload distribution tools like GitOps or
> [Work API](https://github.com/kubernetes-sigs/work-api) can leverage this
> API to make informed decisions about which cluster is best suited to handle
> a particular workload.

Reviewer 6: ```suggestion
In this scenario, a standard API acts as a foundation for multicluster
scheduling decisions across various clusters. This means that the 
API will provide a single point of contact to retrieve information that can be
used to make informed scheduling decisions for workloads across multiple
clusters. For instance, workload distribution tools like GitOps or
[Work API](https://github.com/kubernetes-sigs/work-api) can leverage this
API to make informed decisions about which cluster is best suited to handle
a particular workload.
```

Reviewer 6: As  it reads today sounds like the API can be used for cluster management `a standardized way to distribute workloads across multiple clusters.` that's not true. the `ClusterInventory` provides information, then Schedulers make the decision.

Reviewer 12: AFAIK Kubernetes scheduler uses API fields like taints/tolerations.  I think that a multicluster scheduler will need to leverage API fields like this to make decisions about placement of the workload.

My main thoughts are that if this is a user story as part of this KEP, I think this should be targeted in alpha.  So maybe there should be more thoughts on the API for mutlicluster workload distribution.

Author: @kannon92 We discussed the taint before and finally decide to put it in the beta stage. The reason is taint and tolerations are pairs and it is not easy to shape taint if toleration is missing. Toleration is most likely in another API relating to multicluster scheduling which would be another huge topic. Hence we hope to build this API as common & simple as possible at the first stage, and continuously iterate it. WDYT?

Author: @ArangoGutierrez this story is updated

Reviewer 6: As discussed on the `sig-multicluster` call, a strategy of ` continuously iterate it. WDYT?` is not appropriate for a new API, if the first version is of no use for the community, no one will participate in the conversation moving forward, and the iterations won't be successful.  We know there's a lot of need in the community for an Unified API for MultiCluster inventory, but v1alpha1 should serve a broad set of users from the beginning.

Reviewer 12: > @kannon92 We discussed the taint before and finally decide to put it in the beta stage. The reason is taint and tolerations are pairs and it is not easy to shape taint if toleration is missing. Toleration is most likely in another API relating to multicluster scheduling which would be another huge topic. Hence we hope to build this API as common & simple as possible at the first stage, and continuously iterate it. WDYT?

This sounds good to me. From what I understand, this KEP will be iterated outside of core k8s.  I think I'd like to see high level goals for this project on this KEP as I'm sure the API will be iterated on in the subproject.

This could be based on factors such as the current load on the cluster,
its capacity, and its proximity to the data source or end-users. For
example, if a cluster is already running at high capacity, the API could

> direct the workload to a less busy cluster. This ensures optimal

Reviewer 6: As my comment above, this is not true. the API won't direct workloads. It provides information for Schedulers to make a decision. but the `ClusterInventory` CRD  is not a Scheduler API.  That would be MultiCluster-kueue or similar

Reviewer 13: Thanks for pointing out. You're right. The API could provide useful information for the scheduler to make a precise decision, such as workload failover, re-balancing, etc.

Author: updated

Reviewer 6: lgtm

resource utilization and can significantly improve the performance and
efficiency of applications running on the clusters.

#### Story 2: Operations and Management

For multi-cluster administrators, the unified API provides a
comprehensive view of all the clusters under their management. This
includes verifying their memberships, understanding their status,
capacity, and healthiness.

For instance, the API could provide real-time data about the CPU usage,
memory consumption, network latency, and other key performance
indicators of each cluster. If a cluster is nearing its capacity or
experiencing issues, the API could alert the administrators, allowing
them to take proactive measures to prevent downtime or performance
degradation.

Moreover, the API could also provide insights into the membership of
each cluster, such as the number of nodes, their roles, and their
current status. This can help administrators manage the clusters more
effectively and ensure that they are always operating at optimal levels.

> #### Story 3: Transparent to Consumers

Reviewer 6: This User Story is also a key point on why linking NFD is a good thing for the community.
Currently there are many end-users as well as projects that have NFD as their dependency to operate. E.g is the GPU-Operator from NVIDIA, the Confidentials-Container Operator, The Network-Operator, and many other infrastructure projects, rely on NFD for operating the cluster. So having the same information at the Cluster level, would be of relevance for projects like HyperShift(OpenShift RedHat) were they rely heavily on NFD API's

Reviewer 14: agree on the high level. However, I think we need to adapt the NFD to CFD to work with clusterInventory as we cannot just lay out 5k node features in it. The cluster level scheduler also shouldn't need to know the feature of each node.

Reviewer 15: @ryanzhang-oss that for sure doesn't work. But what do you think about adding a field for per-node features (like `nodeProperties` in addition to properties):

```yaml
nodeProperties:
   - name: node-1
     properties:
       - name: my-property
         value: foobar
```

This could then be used to expose per-node information if/when needed

Reviewer 6: About: 

```
The cluster level scheduler also shouldn't need to know the feature of each node.
```

I am working with multiple stakeholders, not just at my ORG, but in academia and other cloud vendors, where we need this information for the scheduler. When it comes to specialised workloads like HPC (MPI) and AI/ML, the scheduling decision can not be done just with CPU/Memory of the entire cluster

Author: would you give more details on why per node info is needed? Why exposing them as a cluster-wide property is not sufficient for multiple cluster scheduling? Does it also need to select a certain node in the cluster in addition to selecting an appropriate cluster?

Reviewer 6: For some MPI or AI/ML workloads, we need not only to know if the cluster is GPU capable, but if the nodes are connected with fast network, how many nodes share this feature, etc. 
We don't expect this information to be used to decide the final node from the central management cluster, but to make an informed decision on which cluster suits best our workload needs

Author: Aren't these can also be exposed as labels or cluster wide properties, e.g. using `ClusterProperty` API in https://github.com/kubernetes-sigs/about-api? It seems to me these can be defined as "the number of nodes in the cluster having fast network"?

Reviewer 6: lgtm

> #### Story 3: Transparent to Consumers

Author: Updated the terminology section to clarify the consumers

multicluster scheduling, workload distribution and cluster management.
The API aims to provide the information for the consumers to answer the
below questions:

* Is the cluster under management?
* How can I select a cluster?
* Is the cluster healthy?
* Does the cluster have certain capabilities or properties?
* Does the cluster have sufficient resources?

### Cluster Name

It is required that cluster name is unique for each cluster, and it
should also be unique among different providers (cluster manager). There
are two option to ensure it is uniqueness

#### Option 1

The `metadata.name` of the cluster should never be set upon creation.
Only `metadata.generateName` can be set.

#### Option 2

The `metadata.name` and `metadata.generateName` must have prefix which
should be the same as the `spec.clusterManager.name`. Different cluster
manager must set a different value of `spec.clusterManager.name` when
the cluster is created.

These two options can be achieved by a validating webhook or a
`ValidatingAdmissionPolicy` resource.

### Spec

#### Display name

It is a human-readable name of the cluster set by the consumer of the
cluster.

> #### Cluster Manager

Reviewer 16: Why do you think that? 
Do you mean it is redundant with the predefined label `x-k8s.io/cluster-manager`?

> #### Cluster Manager

Reviewer 6: yeah, 1 is redundant, and second, some users might not have a cluster-manager. they can have in-house tools for multi-cluster management

> #### Cluster Manager

Reviewer 4: > yeah, 1 is redundant, and second, some users might not have a cluster-manager. they can have in-house tools for multi-cluster management

It seems pretty reasonable to expect them to name their manager to avoid ambiguity and for their own evolution over time.

> #### Cluster Manager

Reviewer 6: lgtm

An immutable field set by a cluster manager when this cluster resources
is created by it. Each cluster manager instance should set a different
values to this field.

> In addition, a predefined label with key "x-k8s.io/cluster-manager" needs

Reviewer 4: See if you can make a validating admission policy to enforce this. Recall that you can use CRD validation on fields, but not on labels.

Reviewer 13: Sure. CEL expressions could help make sure this field is immutable.

Actually, there are other purposes that we want to have this **OPTIONAL** label.

- With this label, it will be easier for multiple cluster managers to query the clusters and get notified on clusters that they are responsible. Although this is doable with lister indexers/filtering funcs, labels are handy and quite helpful.
- This label could also serve as a default setting of cluster manager.

to be added by the cluster manager upon creation. The value of the label
MUST be the same as the name of the cluster manager. The purpose of this
label is to make filter clusters from different cluster managers eaiser.

> ### Status

Reviewer 6: I disagree with Status, it should be Spec.
as CRD"s usually have a Status field dedicated to event's on the object.
Look at https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition, where `version` is inside `spec` not Status

The status subresource of a CRD can be considered to have the same objective of non-custom resources. While the spec defines the desired state of that particular resource, basically I declare what I want, the status instead explains the current situation of the resource I declared on the cluster and should help in understanding what is different between the desired state and the actual state.

Like a StatefulSet spec could say I want 3 replicas and its status say that right now only 1 of those replicas is ready and the next one is still starting, a custom resource status may tell me what is the current situation of whatever I declared in the specs.

Reviewer 13: Here the version field is kind of static metadata. It is not configurable, which is quite different from those in `spec`.

In this KEP, cluster inventory API is not used to provision a cluster from scratch. That is what `cluster-api` has already done.

Reviewer 6: I am referring to the 'status' field. I know this API is not for provisioning, but for living clusters. Still I think 'status' should not be used as proposed here.

Author: I think the field currently defined in the status is actual state but not desired state. These fields in status is supposed to be updated by the cluster manager but not configured by the consumer.

Reviewer 6: I think my initial comment is miss-understood, My point is that I think `status` field is not being used as the default way CRD's use status. Look at https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition

Reviewer 14: @ArangoGutierrez  I still don't follow. As @qiujian16 and @dixudx explained, the clusterAPI status fields are not desired state of the cluster. Where do you think we should put them then?

Reviewer 6: So currently this proposes:
```yaml
spec:
  displayName: some-cluster
  clusterManager:
    name: some-cluster-manager
status:
 version:
   kubernetes: 1.28.0
 resources: # total resources of a cluster
   capacity:
     cpu: 12
     memory: 12Gi
   allocatable:
     cpu: 10
     memory: 16Gi
```

everything in Status, should be in Spec.

And Status be just for 
```yaml
 conditions:
   - type: ControlPlaneHealthy
     status: True
     lastTransitionTime: "2023-05-08T07:56:55Z"
     message: ""
   - type: AllNodesHealthy
     status: True
     lastTransitionTime: "2023-05-08T07:57:55Z"
     message: ""
   - type: Joined
     status: True
     lastTransitionTime: "2023-05-08T07:58:55Z"
     message: ""
   ```

Reviewer 6: Cluster resources mutate, more in cloud environments were nodes can be added/removed dynamically. In my use case, nodes are added and removed on a daily basis 
Same for Version, what if the cluster if Updated?

Author: that is my point, but this shouldn't be mutated by user, but the cluster manager.

Reviewer 6: Why the cluster manager? resource information, and other features should be advertise at the cluster level.
Question(maybe is not clear to me yet): where do you intend this API to live, on each cluster or in the management cluster?

Author: It is supposed to be live in a management cluster.

Reviewer 6: I see, then my view was wrong.  I will remove some one my comments then

Reviewer 6: lgtm

#### Version

> Kubernetes version of the cluster

Reviewer 13: Yeah, possibly. But we need to consider how to incorporate these versions for hundreds, thousands of nodes in a single cluster. We may have to make a compromise, like exactly how kubelet reports container images in its status.

> Kubernetes version of the cluster

Reviewer 6: I agree, I forgot about heterogeneous clusters, and each node could have a diff version, it gets messy. we can set this threat as solved

> Kubernetes version of the cluster

Reviewer 4: With recent conversations about kube-apiserver and enabled featureset version, I can see trying to accommodate that in an API.  Perhaps a few well-known keys: control-plane, minimum kubelet version, maximum kubelet version, and enabled featureset version.  Enumerating every kubelet version is a losing proposition.

> Kubernetes version of the cluster

Reviewer 6: lgtm

Versions of the kubernetes can let consumers understand the capability
of the kubernetes, such as what API is supported.

#### Resources

Resources include capacity and allocatable, which is the sum of the
capacity/allocatable of nodes on the cluster. It is useful for users to
know the sizing of the cluster, and also schedulers to sort or bin pack
the workload to clusters.

#### Properties

Name/value pairs to represent properties of the clusters. It could be a
collection of ClusterProperty resources, but could also be info based on

> other implementations.

Author: the reason is trying to align with `ClusterProperty` API https://github.com/kubernetes-sigs/about-api. It seems to me it is a different use case to introduce topology information of the cluster.

> other implementations.

Reviewer 6: Is there a reason of why trying to align with that API? Also that was not stated in the description. So maybe more information is needed to convey your idea for the `properties` field.

> other implementations.

Reviewer 8: The purpose is to provide cluster level properties. Checked the https://github.com/kubernetes-sigs/node-feature-discovery/issues/1423 and `features` is also a key-value pair map. So they're aligned? 

```
// Features is the set of cluster wide features that are not reported at the node level.
// +optional
Features map[string]string `json:"features"`
```

> other implementations.

Reviewer 6: Yup, I think NFD https://github.com/kubernetes-sigs/node-feature-discovery/issues/1423 aligns perfectly with this KEP, a few changes, and we could merge both ideas. 
NFD has a big user base requesting for this API at the NFD level, but having it here will make a bigger impact

> other implementations.

Reviewer 6: Also I checked, and an average NFD `NodeFeature` CRD weights around 80k, meaning we could safely host a `ClusterInventory` of 20k nodes without reaching etcd limits of 1.5M
cc @qiujian16

> other implementations.

Reviewer 4: Name-value pairs match well with the ClusterClaim API and the ClusterProperty from discussion with sig-arch a couple years back.

> other implementations.

Reviewer 6: I agree, I am just saying we could have something like
```go
// ClusterFeatureSpec describes a ClusterFeature object.
type ClusterFeatureSpec struct {
	// Nodes is the full "raw" features data per node that has been discovered.
	// +optional
	Nodes []NodeFeatureSpec `json:"nodeFeatureSpec"` <--- from k8s-sig/nfd
	// Features is the set of cluster wide features that are not reported at the node level.
	// +optional
	Features map[string]string `json:"features"`
}
```

> other implementations.

Reviewer 14: @ArangoGutierrez  I wonder why we need to have a list of the node features?

> other implementations.

Reviewer 6: Is an Inventory of the Cluster, for HPC/AI/ML use cases, is necessary to know specific node features to be able to perform a good scheduling decision

> other implementations.

Reviewer 13: > is necessary to know specific node features to be able to perform a good scheduling decision

Then this comes to a question, whether we want to design a single-layer scheduling flow or a normal two-layer scheduling flow.

- single-layer scheduling: The multi-cluster scheduler (running in multi-cluster management control plane) is responsible to perform an **End-to-End** scheduling decision, i.e., selecting the perfect nodes in child/member clusters. In this way, it is true and necessary that all the specific node features in every cluster should be reported. But, please also be aware of the scheduling efficiency, heartbeat burdens, lots of data, scheduling failure, etc.

- two-layer scheduling: In this way, the multi-cluster scheduler is only responsible to select the perfect clusters, instead of nodes. We only need to make sure the target cluster is capable to run the workloads, regardless of detailed/specific node features. Normally what the multi-cluster scheduler really needs is a raw estimation/prediction to prove that cluster has enough resources, node features. With cluster labels indicating specific node features are helpful as well.

IMO, for multi-cluster scheduling, having a per-node data is superfluous. And it is too heavy when we have clusters and nodes in a large scale. I agree with @deads2k that it is better to have some well-known node features or we tweak with some kind of node feature aggregations.

> other implementations.

Reviewer 6: lgtm

#### Conditions

Record cluster’s healthiness condition and easy to extend, conforming to
metav1.Condition format.

Predefined condition types:
- Healthy conditions indicate the cluster is in a good state. (state:
  True/False/Unknown). Healthiness can have different meanings in
  different scenarios. We will have multiple condition types to define
  healthiness:
  - ControlPlaneHealthy is to define whether the controlplane of the
    cluster is in the healthy state.
    * apiserver/healthz
    * controller-manager/healthz
    * scheduler/healthz
    * etcd/healthz
  - AllNodesHealthy is to define if the nodes in the cluster are in a
    healthy state. If one node is not healthy, the status of NodeHealthy

>     is set to false with the message indicating details. (todo

Reviewer 4: use the `<<[UNRESOLVED]>>` marker on this.  I suspect we'll want default tolerances.  large clusters always have downed nodes.

Author: added <<[UNRESOLVED]>> marker

Reviewer 4: To close out this unresolved section, could the AllNodesHealthy be deferred in favor of counts as suggested in https://github.com/kubernetes/enhancements/pull/4323/files#r1382156898 or alternatively, throwing out a concrete suggestion of 95% of nodes healthy to represent a "mostly working" cluster.

    tolerance, it should be configurable)

>   It would be useful to collect the healthiness of other subsystems in

Reviewer 4: One thought, it ought to be possible to show a count of some node markers
1. total nodes
2. number of schedulable nodes
3. number of ready nodes

  the cluster, e.g. network, dns, storage, or ingress. However, it is
  not easy to collect that information in a common way with different
  implementations of network or storage providers. We decide not to
  include other subsystems healthiness conditions in the initial phase.
- Joined: indicate the cluster is under management by the cluster manager.
  The status of the cluster SHOULD be updated by the cluster manager under
  this condition.

## API Example

```yaml
apiVersion: multicluster.x-k8s.io/v1alpha1
kind: ClusterInventory
metadata:
 name: generated-cluster-name
 labels:
   x-k8s.io/cluster-manager: some-cluster-manager
spec:
  displayName: some-cluster
  clusterManager:
    name: some-cluster-manager
status:
 version:
   kubernetes: 1.28.0
 resources: # total resources of a cluster
   capacity:
     cpu: 12
     memory: 12Gi
   allocatable:
     cpu: 10

>      memory: 16Gi

Reviewer 6: The way this resources are presented are of little help for a resource manager/scheduler.  12 cpu's could mean a node with 10 an another with 2. Then is impossible to make an informed scheduling decision based on the presented info

Author: yes, currently it is the sum of all nodes in a cluster to let consumer understand the overall capacity. Agree it is not quite helpful for scheduler. But resources has a map structure that it can also add info like `max_cpu_per_node` and `min_cpu_per_node`. My understanding is the scheduler on the multicluster level is to select a cluster, while the node selection is still done by the kube-scheduler, so the above info (`max_cpu_per_node`) would be enough for multicluster scheduler?

Reviewer 6: well `max_cpu_per_node` can be confusing, if you have one "big" node and a bunch of small ones. 
That's why this conversation, I think is the most important on this review https://github.com/kubernetes/enhancements/pull/4323#discussion_r1377254904

Reviewer 4: > well `max_cpu_per_node` can be confusing, if you have one "big" node and a bunch of small ones.

I think it could be reasonable to expect different "classes" of clusters to have different ClusterProperties to allow selective targeting if needed.  ARM versus AMD is another case of "nodes aren't the same", but if you're multi-cluster first, it's likely that clusters could be labelled in meaningful ways.

Reviewer 6: lgtm

- A dummy controller and unit test to validate the CRD and client.

#### Beta

- Gather feedback from users during the Alpha stage to identify any
  issues, limitations, or areas for improvement. Address this feedback
  by making the necessary changes to the API and iterating on its design
  and functionality.
- The API should support the addition of scheduling features, such as:
  - Load balancing: Distribute workloads evenly across clusters based on
    their current load and capacity.
  - Affinity and anti-affinity rules: Allow users to define rules for
    placing workloads on specific clusters or ensuring that certain
    workloads are not placed on the same cluster.
  - Priority-based scheduling: Enable users to assign priorities to
    workloads, ensuring that higher-priority workloads are scheduled
    before lower-priority ones.
  - Resource-based scheduling: Schedule workloads based on the
    availability of specific resources, such as CPU, memory, or storage,
    in each cluster.
- The API should expose access information including but not limited to:
  - APIServer endpoint url of the member cluster.
  - Credential with limited access to the member cluster.

> - At least two providers and one consumers using cluster inventory API.

Reviewer 4: +1 for this.
